### PR TITLE
feat: Revamp of product categories

### PR DESCRIPTION
**Pull request by sradevski**
There are several changes, namely:
1. Refactored the product category interface to follow the general module standards
2. Refactored the product category repository to be much easier to understand
3. Fixes CORE-2287
4. Minor fixes to how ranks were normalized

### CODE REVIEW

File: packages/modules/product/src/models/product-category.ts [hunk 1]
```diff
@@ -130,22 +130,7 @@ class ProductCategory {
...
-        this.parent_category_id
-      )
-    }
-
-    if (parentCategory) {
-      this.mpath = `${parentCategory?.mpath}${this.id}.`
-    } else {
-      this.mpath = `${this.id}.`
-    }
+    this.mpath = `${this.mpath ? this.mpath + "." : ""}${this.id}`
```
Comment (carlos-r-l-rodrigues) [inline packages/modules/product/src/models/product-category.ts:133]: if updating the `parent_id`, this will get the incorrect reference no?

Comment (carlos-r-l-rodrigues) [reply packages/modules/product/src/models/product-category.ts:133]: nevermind. just checked the repository.

**Review by carlos-r-l-rodrigues — APPROVED** [review]

**Review by olivermrbl — COMMENTED** [review]
Very nice work and clean-up!

I think we still have issues with the ranking though. I captured some of the behaviour I was seeing below:

https://github.com/medusajs/medusa/assets/59018053/af5b78ed-1743-4add-a2d2-5817d705065f

**Comment by sradevski** [comment]
@olivermrbl yes I noticed the same thing, I already messaged Kasper in Slack. The rank number is correctly updated as you can see, so it could be a UI bug. If not, there might be something not working correctly in the `find` method.

I just checked, `find` doesn't do any rank sorting currently. I'll do it in a separate PR, this one is big enough as it is.

**Pull request merged**